File: Progs/Random/MergeSort.java
import java.util.*;

class MergeSort {
    public static void main(String[] args) {
	int[] array = {};//{34,5,3,3,5,3234,24,53};
	new MergeSort().mergeSort(array);
	System.out.println(Arrays.toString(array));
    }

    void mergeSort(int[] array) {
	int length = array.length;
	mergeSort(array, 0, array.length - 1);
    }

    void mergeSort(int[] array, int start, int end) {
	if (start < end) {
	    int mid = ((start + end) / 2) + 1;
	    mergeSort(array, start, mid - 1);
	    mergeSort(array, mid, end);
	    merge(array, start, mid, end);
	}
    }

    void merge(int[] array, int start, int mid, int end) {
	int[] temp = new int[end - start + 1];

	int i = start;
	int j = mid;
	int c = 0;
	
	while(i < mid || j <= end) {
	    if (i >= mid) {
		temp[c] = array[j++];
	    }
	    else if (j > end) {
		temp[c] = array[i++];
	    }
	    else if (array[i] >= array[j]) {
		temp[c] = array[i++];
	    }
	    else {
		temp[c] = array[j++];
	    }
	    c++;
	}
	
	for (c = 0; c < temp.length; c++) {
	    array[start++] = temp[c];
	}
    }
}

File: Progs/Random/mergesort.cpp
#include <iostream>

using namespace std;

void merge(int *a, int low, int mid, int up)
{
 int p1 = low, i;
 int p2 = mid+1;
 int b[100];
 cout << low << " " << mid << " " << up << " " << endl;
 for(i=low; i<=up; i++)
	{
	 if ((a[p1] < a[p2] && p1<=mid) || p2>up)
		b[i] = a[p1++];
	 else if(p2<=up)
		b[i] = a[p2++];
	}
 for(i=low; i<=up; i++)
	{//cout << i << " " << b[i] << endl;
	 a[i]=b[i];
	}
 cout << endl << endl;
}
void mergesort(int *a, int low, int up)
{int  	mid = (low+up)/2;
 if(up>low)
	{
	 //mid=(up+low)/2;
	 mergesort(a,low,mid);
	 mergesort(a, mid+1, up);
	 merge(a, low, mid, up);
	}
}
int main()
{
 int a[]={34,54,7,2,5,76,24,34,6,75}, i;
 mergesort(a, 0, 9);
 for(i=0; i<10; i++)
	cout << a[i] << endl;
}
 